### ISSUE TITLE
Questions regarding failed evaluations and dynamic updates of the surrogate function

### ISSUE THREAD

**Issue by TimoSommer** [ARM-software/mango#33]
Hey,
I have some questions regarding the documentation, it would be great if you could update it. Additionally I have one feature request, depending on how Mango works at the moment.

First, I'm interested in how exactly Mango handles failed evaluations. Neither the documentation nor the paper are really clear about this. From my tests, I can see that if I have some jobs failing I effectively get more runs than without jobs failing. But it looks like this is only the case for the initial random runs, not for the Bayesian runs. And it seems to only try twice until it gives up, is that true? It would be nice if you could add a bit to the documentation about this.

Second, I wonder if it is possible to use the mechanism of failed evaluations to somehow have a dynamic update of the surrogate function. In my application the evaluations can take quite different runtimes depending on the hyperparameters. Therefore it would be cool if I could update the surrogate function every time I get one new result, this is what I mean with a dynamic update of the surrogate function. Currently I need to wait for the whole batch of hyperparameters to finish its evaluation and only then I can return it.
If possible it would be amazing if you could provide a builtin functionality for this. Otherwise I was wondering if I could use the mechanism of failed evaluations to do this by hand:
1. Get a batch of hyperparameters. Evaluate them in parallel by submitting a new task for each (on Slurm).
2. Whenever a single task returns the score, return this to the Mango Optimizer. For the Mango Optimizer, it would look like all but this one evaluations failed.
3. Mango gets me a new batch of hyperparameters, I again submit the Slurm tasks to calculate them. 
4. While waiting for them to return, one of the tasks of the first batch returns its score. I again return this score to the Mango optimizer which will in turn provide me with a new batch of hyperparameters...

I was hoping this would be possible, but since I am not sure about how Mango handles the failed evaluations I'm not sure it works. I feel like it could be an issue because Mango needs to know that the other jobs didn't fail but that they are still in progress, otherwise it might just try them again. I'm very interested to hear your opinion about this.

Also, thanks for this great package! It's really cool, I like it a lot.
Cheers,
Timo

**Comment by sandeep-iitr** [comment]
Hi,

Take a look at this example: https://github.com/ARM-software/mango/blob/master/examples/Failure_Handling.ipynb

This example shows when random failures happen you can dynamically return only successful results. It is also possible to return with only partial completed runs, and you don't have to wait for the entire batch to complete.

**Comment by TimoSommer** [comment]
Thanks for your answer! So if I choose to return only part of the batch, than it is possible to return the other part of the batch later right? But I guess the Mango tuner would have already spawned new jobs in the neighborhood of the jobs I didn't return, because it assumes they failed. Is this correct? It's not a perfect solution but I guess its hard to come up with a tuner that is able to do these dynamical updates and remember which will still come.

**Comment by sandeep-iitr** [comment]
Yes, you can return the other part of the batch later. These results later also will help to improve the internal approximation (surrogate).

It is not necessarily true that new jobs will be spawned in the neighborhood, with each update even with a partial batch, the approximator will change and the next jobs will be the best possible regions.  It is also possible that the next jobs are in the neighborhood. Internally, Mango doesn't remember which jobs are failed. This functionality can be tracked in the scheduler and was kept outside of Mango as detecting failures and keep track of them is often application dependent.

**Comment by TimoSommer** [comment]
Alright, thanks for your extensive answer, you helped me a lot!
